tensorboard/backend/event_processing/data_ingester_test.py

```diff
@@ -251,5 +252,120 @@ def testSingleLetterGroup(self):
+            with mock.patch(
+                "builtins.__import__", side_effect=mock.MagicMock()
+            ) as mock_import:
+                with mock.patch.object(
+                    gfile_mock,
+                    "get_registered_schemes",
+                    return_value=["gs", "s3"],
+                ) as mock_get_registered_schemes:
+                    data_ingester._try_to_support_tfio(
+                        ["tmp/demo", "s3://bucket/123"]
+                    )
+        mock_import.assert_not_called()
+        mock_get_registered_schemes.assert_called_once()
+
+    def testTryToSupportTfio_import(self):
+        with mock.patch.object(tf.io, "gfile") as gfile_mock:
+            gfile_mock.return_value.get_registered_schemes = mock.MagicMock()
+            with mock.patch(
+                "builtins.__import__", side_effect=mock.MagicMock()
+            ) as mock_import:
+                with mock.patch.object(
+                    gfile_mock,
+                    "get_registered_schemes",
+                    return_value=["file", ""],
+                ) as mock_get_registered_schemes:
+                    data_ingester._try_to_support_tfio(
+                        ["tmp/demo", "s3://bucket/123"]
+                    )
+        self.assertEqual("tensorflow_io", mock_import.call_args[0][0])
+        mock_get_registered_schemes.assert_called_once()
+
+    def testTryToSupportTfio_raiseError(self):
+        with mock.patch.object(tf.io, "gfile") as gfile_mock:
+            gfile_mock.return_value.get_registered_schemes = mock.MagicMock()
+            with mock.patch(
+                "builtins.__import__",
+                side_effect=ImportError,
+            ) as mock_import:
+                with mock.patch.object(
+                    gfile_mock,
+                    "get_registered_schemes",
+                    return_value=["file", "ram"],
+                ) as mock_get_registered_schemes:
+                    err_msg = (
+                        "Error: Unsupported filename scheme 's3' (supported schemes: ['file', 'ram'])."
+                        + " For additional filesystem support, consider installing TensorFlow I/O"
+                        + " (https://www.tensorflow.org/io) via `pip install tensorflow-io`."
+                    )
+                    with self.assertRaisesWithLiteralMatch(
+                        tf.errors.UnimplementedError, err_msg
+                    ):
+                        data_ingester._try_to_support_tfio(
+                            ["tmp/demo", "s3://bucket/123"]
+                        )
+        self.assertEqual("tensorflow_io", mock_import.call_args[0][0])
+        mock_get_registered_schemes.assert_called_once()
+
+    def testTryToSupportTfio_fallback(self):
+        with mock.patch.object(
+            tf.io.gfile, "get_registered_schemes"
+        ) as mock_get_registered_schemes:
+            mock_get_registered_schemes.return_value = None
```

Comment:
It's not quite accurate to mock get_registered_schemes() as returning None, since that isn't what would actually happen with the real code (it just happens to also trigger the fallback). The actual scenario is that the symbol isn't present at all. I think we could simulate that by doing:

```python
with mock.patch.object(tf.io, "gfile", autospec=True) as mock_gfile:
   del mock_gfile.get_registered_schemes
```

Also, as a nice bonus, if you have `mock_gfile`, you should be able to just do `mock_gfile.exists.return_value = True` below, rather than needing to call `mock.patch.object()` again separately.

Reply:
Thanks for the suggestion!